### PR TITLE
[Debug view] Add dynamic debug configurations

### PR DESCRIPTION
**Pull request by alvsan09**
#### What it does

Following the support for dynamic debug configurations via
command Debug: Select and Start debugging. PR #10134

This change extends the functionality further to bring provided
dynamic debug configurations to the list of available configurations
under the debug view.

The user can select these dynamic configurations by first selecting
the debug type and a quick pick input box will follow to allow the
selection from the available configurations of the selected type.

Once a dynamic configuration has been selected it will be added to a
short list (limited to 3) of recently used dynamic debug
configurations, these entries are available to the user, rendered
with the name and suffixed with the debug type in parenthesis.
This will facilitate subsequent selection / execution.

This change additionally preserves and restores the list of recently
selected  dynamic debug configurations so they are presented in
subsequent sessions.

These configurations are refreshed from the providers when:
   - Configuration providers are registered or unregistered
   - Focus is gained or lost by the configuration selection box

Refreshing these configurations intends to render valid dynamic
configurations to the current context.
e.g.
   - Honoring an extension 'when' clause to a file with a specific
     extension opened in the active editor.

However there are situations where the context for execution of a
dynamic configuration may no longer be valid e.g.
   - The configuration is restored from a previous session and the
     currently selected file is not supported.
   - The switch of context may not have involved a refresh of dynamic
     debug configurations. (e.g. switching active editors, etc.)
Considering the above, execution of dynamic configurations triggers
the fetch of dynamic configurations for the selected provider type,
if the configuration is no longer provided, the user will be notified
of a missing configuration or not applicable to the current context.

It additionally
Fixes: #11050
The update were required to integrate  the solution with the `SelectComponent`

#### How to test

1. Assuming the extension `ms-vscode.js-debug` is installed as a builtin dependency
Open a `javascript` or `typescript` project, build it  and open in editor a `.js` or `.ts` file ready to be debugged.
`Open the Debug view and open the selection box.`
2. Select the entry with an ellipsis suffix (...) after a separator e.g.  `pwa-node...`
3. A quick pick will appear with the dynamic configurations provided, Select the one called `Run current file` and start it.
4. Verify that the debug session starts and that debug options work (e.g. stop at break points, etc.).
5. End the debug session,
6. Test that other non dynamic configurations are working as usual.
7. Verify that after selecting a dynamic configuration it becomes directly available under the main selection box
8. Restart the application and verify that the recently used dynamic configuration become available directly under the selection box.
9. Verify that a dynamic configuration selected in the main box can be restored after restarting the application.
10. Repeat the tests above using a `multi-root` workspace.

#To test with additional dynamic configurations, you can install a modified 'vscode-mock-debug' extension updated
to test `Theia`:
https://github.com/alvsan09/vscode-mock-debug/releases/download/v1.18-alpha/mock-debug-theia-0.46.3.vsix

The 'vscode-mock-debug' extension is meant to simulate debug on `.md` files although you can also try it on other types.
dynamic_config_view gif

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by alvsan09** [comment]
This PR is converted to draft in order to addressed comments raised by Paul in offline discussion.
i.e.
* Update the debug configuration manager to avoid a cache
* Dynamic configurations shall be resolved on the spot, if they are restored from a previous session and are no longer valid, a message shall be given to the user.
* The above to follow vscode intended implementation and to reduce complexity.

File: packages/debug/src/browser/view/debug-configuration-select.tsx
Comment (colin-grant-work) [inline packages/debug/src/browser/view/debug-configuration-select.tsx:24]: The `I` prefix [is discouraged](https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#interfaces-no-i-prefix)
```suggestion
interface DynamicPickItem { label: string, configurationType: string, request: string, providerType: string }
```

File: packages/debug/src/browser/view/debug-configuration-select.tsx
Comment (colin-grant-work) [inline packages/debug/src/browser/view/debug-configuration-select.tsx:159]: It's usually discourage to use indices as keys in React if it can be avoided. Is there a way to create a unique, persistent identifier from the information used to create the `option`?

File: packages/debug/src/browser/view/debug-configuration-select.tsx
Comment (colin-grant-work) [inline packages/debug/src/browser/view/debug-configuration-select.tsx:39]: the `NO_CONF` and `ADD_CONF` constants could be similarly extracted.

**Review by colin-grant-work — COMMENTED** [review]
Some minor comments on the code.

File: packages/debug/src/browser/debug-configuration-manager.ts
Comment (alvsan09) [reply packages/debug/src/browser/debug-configuration-manager.ts:184]: We can not resolve 'name' this way as it comes from the `configuration` property

Comment (colin-grant-work) [reply packages/debug/src/browser/debug-configuration-manager.ts:184]: Good point. My eye skipped over that part. Could do `{ providerType, configuration: { name }}`, if you'd like, but the nested destructuring can get a bit hard to read.

Comment (alvsan09) [reply packages/debug/src/browser/debug-configuration-manager.ts]: your suggestion looks fine to me, it's now included in the follow up commit!

File: packages/debug/src/browser/view/debug-configuration-select.tsx
Comment (vince-fugnitto) [inline packages/debug/src/browser/view/debug-configuration-select.tsx:105]: The following method does a lot, perhaps it should be broken down?

File: packages/debug/src/browser/view/debug-configuration-select.tsx
Comment (vince-fugnitto) [inline packages/debug/src/browser/view/debug-configuration-select.tsx:140]: I believe `this.quickInputService.showQuickPick` can replace all of this code.
It accepts the items, and supports configuration options such as `placeholder`.

File: packages/debug/src/browser/debug-configuration-manager.ts
Comment (alvsan09) [reply packages/debug/src/browser/debug-configuration-manager.ts:213]: There are only two places where we add options to the tracker 
1) With the `unshift` above
2) When restoring options saved in a previous session

So practically only 1) is relevant (unless local storage is manipulated), so we are checking directly after each addition, 
since we add items to the tracker one by one, I don't see how we could go higher than the hard coded limit. 
If you could please expand on the scenario.

Comment (vince-fugnitto) [reply packages/debug/src/browser/debug-configuration-manager.ts:213]: @alvsan09 I just prefer more defensive coding that would also be clearer. The goal is to have 3 elements as not to clutter the dropdown, but the current code does not guarantee it, for example if in the future there are additional callers that add additional items to the list. (I believe it is why vscode uses `slice`) 

I'll leave the decision up to you however! :)

Comment (paul-marechal) [reply packages/debug/src/browser/debug-configuration-manager.ts:213]: I like the idea of using slice! As Vincent said this will ensure that the list always has the expected length.

File: packages/debug/src/browser/debug-configuration-manager.ts
Comment (paul-marechal) [reply packages/debug/src/browser/debug-configuration-manager.ts:466]: You can also just skip the `const` definition and directly define it as the argument passed to `push`.

File: packages/debug/src/browser/debug-configuration-manager.ts
Comment (alvsan09) [reply packages/debug/src/browser/debug-configuration-manager.ts]: I per our discussion (off-line) I have replaced `DebugSessionOptionsData` with `DebugSessionOptions` and reflected in the latest commit.

File: packages/debug/src/browser/debug-configuration-manager.ts
Comment (alvsan09) [reply packages/debug/src/browser/debug-configuration-manager.ts]: I agree: 
The new code uses `splice` as follows: 
                this.recentDynamicOptionsTracker.splice(recentMax);

File: packages/debug/src/browser/debug-configuration-manager.ts
Comment (alvsan09) [reply packages/debug/src/browser/debug-configuration-manager.ts]: This code is no longer used as `DebugSessionOptionsData` has been replaced by `DebugSessionOptions`

File: packages/debug/src/browser/view/debug-configuration-select.tsx
Comment (alvsan09) [reply packages/debug/src/browser/view/debug-configuration-select.tsx]: Done! Thanks !

File: packages/debug/src/browser/debug-configuration-manager.ts
Comment (paul-marechal) [inline packages/debug/src/browser/debug-configuration-manager.ts:217]: Let's trust our users and just type the return as `readonly DebugSessionOptions[]`:

```suggestion
    get recentDynamicOptions(): readonly DebugSessionOptions[] {
        return this.recentDynamicOptionsTracker;
    }
```

**Review by vince-fugnitto — COMMENTED** [review]
The code and functionality works well for me :+1: 
@colin-grant-work do you mind doing a review pass as well when you get a chance?

File: packages/debug/src/browser/view/debug-configuration-select.tsx
Comment (colin-grant-work) [inline packages/debug/src/browser/view/debug-configuration-select.tsx:120]: Matter of taste only:
```suggestion
        if (!configurationsOfProviderType) {
            return [];
        }

        return configurationsOfProviderType.configurations.map(configuration => ({
            label: configuration.name,
            configurationType: configuration.type,
            request: configuration.request,
            providerType
        }));
```

File: packages/debug/src/browser/view/debug-configuration-select.tsx
Comment (colin-grant-work) [inline packages/debug/src/browser/view/debug-configuration-select.tsx:36]: I don't think that it is very likely that this particular element will be overridden downstream, but in principle an injectable `ReactRenderer` would make the task easier for adopters.

File: packages/debug/src/browser/view/debug-configuration-select.tsx
Comment (colin-grant-work) [inline packages/debug/src/browser/view/debug-configuration-select.tsx:175]: I believe this should be conditional on the presence of non-dynamic options. Otherwise it's possible to end up with an initial separator:

image

Unrelated to this PR, the styling of our `select` elements is pretty bad on MacOS (screenshot from Chrome).  I'll make an issue for that.

**Review by colin-grant-work — CHANGES_REQUESTED** [review]
Some minor comments on the code below. In addition, it's possible to get into a weird state if a dynamic debug configuration is the only kind available.

> 1. Select the entry with an ellipsis suffix (...) after a separator e.g. pwa-node...
> 2. A quick pick will appear with the dynamic configurations provided, Select the one called Run current file and start it.

If a dynamic configuration provider is the only option available, then it is the default selection. That means that 'selecting' it again does not trigger a change event, so it's impossible to get the desired behavior without creating some other kind of configuration and selecting that so that you can then select the dynamic provider _and_ trigger a change event. In VSCode under similar circumstances, a welcome view is displayed that includes a button to reveal automatic debug configs, but within the bounds of our current UI, probably we should just ensure that a dynamic provider is never actually the selected item.

File: packages/debug/src/browser/debug-session-options.ts
Comment (paul-marechal) [reply packages/debug/src/browser/debug-session-options.ts:48]: I got confused by your comment, but I think you are right and we should return an object with the breakdown as keys.

**Comment by alvsan09** [comment]
> If a dynamic configuration provider is the only option available, then it is the default selection. That means that 'selecting' it again does not trigger a change event, so it's impossible to get the desired behavior without creating some other kind of configuration and selecting that so that you can then select the dynamic provider _and_ trigger a change event. In VSCode under similar circumstances, a welcome view is displayed that includes a button to reveal automatic debug configs, but within the bounds of our current UI, probably we should just ensure that a dynamic provider is never actually the selected item.

Thanks for bringing this up, it was a case I did not see,
The 'No configurations' entry shall be added if no configurations are present in the `launch.json` and no recent configurations are available.  I have updated it accordingly in the latest commit.

File: packages/debug/src/browser/view/debug-configuration-select.tsx
Comment (alvsan09) [reply packages/debug/src/browser/view/debug-configuration-select.tsx:36]: As per our discussion, lets keep this in mind for a future enhancement, when needed !
Thanks !!

**Review by colin-grant-work — COMMENTED** [review]
My previous comments have all been addressed. Thanks! 👍

However, when I tested this, I got a very strange result. I'm not sure whether it's a problem with code from this PR or a problem with our interaction with the plugin supplying the debug config, but the following happened:

 - open a folder containing some JS files
 - open the debug view
 - use the picker to select pwa-node...
 - select the 'Create JavaScript debug terminal' option from the quick pick
 - Click play button in debug view
 - --> numerous debug terminals were spawned one after another with no further interaction on my part.

**Comment by alvsan09** [comment]
> However, when I tested this, I got a very strange result. I'm not sure whether it's a problem with code from this PR or a problem with our interaction with the plugin supplying the debug config, but the following happened:
> 
> * open a folder containing some JS files
> * open the debug view
> * use the picker to select pwa-node...
> * select the 'Create JavaScript debug terminal' option from the quick pick
> * Click play button in debug view
> * --> numerous debug terminals were spawned one after another with no further interaction on my part.

This is a separate issue uncovered when first introducing the support of dynamic debug configurations via command,
the following issue was raised back then, see:  
  https://github.com/eclipse-theia/theia/issues/9978

Thanks Colin!

**Comment by colin-grant-work** [comment]
> This is a separate issue uncovered when first introducing the support of dynamic debug configurations via command, the following issue was raised back then, see: #9978
> 
> Thanks Colin!

@alvsan09, is there any chance of #9978 being fixed? Even if the behavior is not the fault of this PR, I don't like the idea of  making it easier for the user to encounter broken things.

**Comment by alvsan09** [comment]
> @alvsan09, is there any chance of #9978 being fixed? Even if the behavior is not the fault of this PR, I don't like the idea of making it easier for the user to encounter broken things.

I have mixed fieelings on this one, I for sure would like to fix them both.

There are several factors to consider:

* The possibility to encounter this issue is already present on master,
the PR that uncovered this issue is already merged:
e.g. See https://github.com/eclipse-theia/theia/pull/10134

* The issue #9978 may not be a trivial fix and may put on hold most of the benefits of this PR while waiting for
its fix.

* If we consider incremental development we can consider this PR as an increment,
  I would vote to merge increments that bring value and cause no actual regressions.
  Note: I don't cosider the uncovering of this issue a regression.

We will eventually fix #9978, and actually I will take a deeper look to it soon,
but given the above arguments I woould vote to merge this PR.

Comments are very welcome!

**Comment by colin-grant-work** [comment]
> Hi @colin-grant-work , From the two latest comments on issue #9978 Would you agree to proceed with this PR first and then perform the refactoring and fixes on top of it?

@alvsan09, I'm still not comfortable approving this PR when it exposes broken behavior in the scenarios I have tested it in. Even if the broken behavior is not a consequence of this change, exposing that broken behavior in a new context is not a positive change.

**Comment by alvsan09** [comment]
Hi @colin-grant-work, the issue #9978 is now fixed and integrated in master. 
This PR has now been uplifted accordingly, 
If you could please re-review this updated PR.

**Review by colin-grant-work — COMMENTED** [review]
This is looking great. There is still one small quirk. In a workspace with just a JS file, if you
1. select `pwa-node...` from the drop down
2. select `JavaScript Debug Terminal` from the options
3. Click the 'Play' arrow to start the session
4. Two debug terminals are created with slightly different labels:

image

They both appear to be hooked up to the debug service, and running the file in either of them will hit breakpoints, but I believe that only one terminal should be created, so we may still be running into some (much more minor) issue in resolving and starting sessions.

**Comment by alvsan09** [comment]
@msujew It would be great to have your opinion on this change, It does include some modifications to the `SelectComponent`

File: packages/core/src/browser/widgets/select-component.tsx
Comment (colin-grant-work) [inline packages/core/src/browser/widgets/select-component.tsx:94]: This class is a bit of an odd duck: it's a React component, but it has also been designed to work in non-React contexts (`PreferenceSelectInput`). Because of that, it's been designed to behave more like an unmanaged input rather than a managed input, to behave more like a vanilla DOM input rather than a typical React input. And I believe that is correct - for it to be useable outside React contexts, it _must_ behave more like an unmanaged input, and be built for use with `Refs`. Given that, its state should in general not be managed by changing `props` but imperatively; the `value` field should be treated as (and perhaps renamed to) `defaultValue` to indicate that it will take effect on initialization, but be ignored thereafter, and `selection` should be moved back into the `state` object: the actual selection should be managed by the component itself, and only updated externally imperatively. I believe moving `selection` from state to a class member is currently only working because every time it is set, it happens to cooccur with a different, React-y means of changing state: at construction time, in `willReceiveProps`, or coincidentally when the value of `state.hover` is also changed.

**Review by colin-grant-work — COMMENTED** [review]
I think the changes to the `SelectComponent` go in the wrong direction: it's intended as a utility for use in both React and non-React contexts, so its lifecycle and data management are a bit non-React-y by design, and changes to make it conform more closely to React norms aren't necessarily desirable. I think it's good to implement additional callbacks and better handling of separators, but the changes to the handling of `state` are working against the component's design.

File: packages/core/src/browser/widgets/select-component.tsx
Comment (alvsan09) [reply packages/core/src/browser/widgets/select-component.tsx]: As mentioned off-line, adding those constraints would make the solution more complicated, 
I have however replaced the `while` statement for a `do-while` statement, to simplify the loop condition.

File: packages/core/src/browser/widgets/select-component.tsx
Comment (alvsan09) [reply packages/core/src/browser/widgets/select-component.tsx]: I have have updated the handling of `state` as per your proposal, 
I just wanted to clarify that the `willReceiveProps` does not change the state,
it simply anticipates the reception of new props passed on from the parent component which then get to be used in the following call i.e. `render`

**Review by colin-grant-work — APPROVED** [review]
Everything is now working for me as indicated. Thanks for your patience @alvsan09!

**Review by msujew — APPROVED** [review]
I'm happy with the changes to the select component. I also did a surface level review of the actual change and everything seems to look good to me.

I just have a minor nitpicky comment, see below.

File: packages/core/src/browser/widgets/select-component.tsx
Comment (alvsan09) [reply packages/core/src/browser/widgets/select-component.tsx]: @msujew, I like your proposal, the code was updated accordingly,  Thanks !!

**Comment by alvsan09** [comment]
> Everything is now working for me as indicated. Thanks for your patience @alvsan09!
@colin-grant-work, thank you Colin!

**Comment by Tesbarmix1** [comment]
> #### What it does
> 
> 
> Following the support for dynamic debug configurations via
> command Debug: Select and Start debugging. PR #10134
> 
> This change extends the functionality further to bring provided
> dynamic debug configurations to the list of available configurations
> under the debug view.
> 
> The user can select these dynamic configurations by first selecting
> the debug type and a quick pick input box will follow to allow the
> selection from the available configurations of the selected type.
> 
> Once a dynamic configuration has been selected it will be added to a
> short list (limited to 3) of recently used dynamic debug
> configurations, these entries are available to the user, rendered
> with the name and suffixed with the debug type in parenthesis.
> This will facilitate subsequent selection / execution.
> 
> This change additionally preserves and restores the list of recently
> selected  dynamic debug configurations so they are presented in
> subsequent sessions.
> 
> These configurations are refreshed from the providers when:
>    - Configuration providers are registered or unregistered
>    - Focus is gained or lost by the configuration selection box
> 
> Refreshing these configurations intends to render valid dynamic
> configurations to the current context.
> e.g.
>    - Honoring an extension 'when' clause to a file with a specific
>      extension opened in the active editor.
> 
> However there are situations where the context for execution of a
> dynamic configuration may no longer be valid e.g.
>    - The configuration is restored from a previous session and the
>      currently selected file is not supported.
>    - The switch of context may not have involved a refresh of dynamic
>      debug configurations. (e.g. switching active editors, etc.)
> Considering the above, execution of dynamic configurations triggers
> the fetch of dynamic configurations for the selected provider type,
> if the configuration is no longer provided, the user will be notified
> of a missing configuration or not applicable to the current context.
> 
> It additionally
> Fixes: #11050
> The update were required to integrate  the solution with the `SelectComponent`
> 
> 
> 
> #### How to test
> 
> 1. Assuming the extension `ms-vscode.js-debug` is installed as a builtin dependency
> Open a `javascript` or `typescript` project, build it  and open in editor a `.js` or `.ts` file ready to be debugged.
> `Open the Debug view and open the selection box.`
> 2. Select the entry with an ellipsis suffix (...) after a separator e.g.  `pwa-node...`
> 3. A quick pick will appear with the dynamic configurations provided, Select the one called `Run current file` and start it.
> 4. Verify that the debug session starts and that debug options work (e.g. stop at break points, etc.).
> 5. End the debug session,
> 6. Test that other non dynamic configurations are working as usual.
> 7. Verify that after selecting a dynamic configuration it becomes directly available under the main selection box
> 8. Restart the application and verify that the recently used dynamic configuration become available directly under the selection box.
> 9. Verify that a dynamic configuration selected in the main box can be restored after restarting the application.
> 10. Repeat the tests above using a `multi-root` workspace.
> 
> #To test with additional dynamic configurations, you can install a modified 'vscode-mock-debug' extension updated
> to test `Theia`:
> https://github.com/alvsan09/vscode-mock-debug/releases/download/v1.18-alpha/mock-debug-theia-0.46.3.vsix
> 
> The 'vscode-mock-debug' extension is meant to simulate debug on `.md` files although you can also try it on other types.
> dynamic_config_view gif
> 
> 
> #### Review checklist
> 
> - [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
> 
> #### Reminder for reviewers
> 
> - As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
>

**Comment by EvilBeaver** [comment]
Excuse me. How to use this functionality from debug extension? I'm an author of Debug Adapter and I want to provide some dynamic debug configurations. What should I do?

**Comment by paul-marechal** [comment]
@EvilBeaver This PR adds support for the `vscode.debug.registerDebugConfigurationProvider` API: use this function from a VS Code extension to contribute debug configurations dynamically.

**Comment by alvsan09** [comment]
@EvilBeaver , 
In the test instructions of this PR, I have used a `mock` debug extension, where you find a good example on how to register your own  dynamic debug configuration provider, the source code  is located at:  
https://github.com/microsoft/vscode-mock-debug

to work better in `theia`, add the following modifications as per the image
image

**Comment by EvilBeaver** [comment]
@alvsan09 thanks for the answer. Probably, Perhaps I was unclear. I'm interested in the part of the functionality called "Dynamic debug configurations" (https://github.com/eclipse-theia/theia/blob/0caea2c27ae5b033751ef7cc041ec9ae953b0bc7/packages/debug/src/node/debug-service-impl.ts#L70)

This is not a part of VSCode API, and all what I could found is a "TODO" stub.

**Comment by alvsan09** [comment]
@EvilBeaver , 
That section of the API is to implement support for `Dynamic Debug configurations` from a `Theia` extension, 
I have not looked into it, and therefore marked as TODO.

However, the functionality is supported via `vscode` extensions using the `vscode` API, would not this be sufficient for your case?
Contributions are very welcome !!

**Comment by EvilBeaver** [comment]
Ок, thanks, now I got it

**Pull request merged**
